### PR TITLE
Port to 1.20.1

### LINKED ISSUE #79: 1.20.1 port
See #78 for more info.

### PR DESCRIPTION
**Pull request by Jab125**

### CODE REVIEW

File: common/src/main/java/com/ultreon/devices/DeviceTab.java [hunk 1]
```diff
@@ -12,9 +15,9 @@
 
 public class DeviceTab {
     @SuppressWarnings("UnstableApiUsage")
-    public static CreativeTabRegistry.TabSupplier create() {
+    public static DeferredSupplier<CreativeModeTab> create() {
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/DeviceTab.java:18]: Not registered correctly yet

File: common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java [hunk 2]
```diff
@@ -3,65 +3,66 @@
...
+import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.item.ItemStack;
 
 import org.jetbrains.annotations.Nullable;
 import java.awt.*;
 
 @SuppressWarnings("unused")
 public class RenderUtil {
-    public static void renderItem(int x, int y, ItemStack stack, boolean overlay) {
+    public static void renderItem(GuiGraphics graphics, int x, int y, ItemStack stack, boolean overlay) {
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java:21]: Unintentional break of method's specification:
1.19.3: Not affected by `PoseStack`
1.20.1: Method now takes `GuiGraphics` which uses a `PoseStack` internally.
A workaround solution is to pop the stack so there are no transformations, and restore it back afterward.

File: common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java [hunk 3]
```diff
@@ -3,65 +3,66 @@
...
     }
 
     /**
      * Texture size must be 256x256
      *
-     * @param pose          the pose stack to draw on
+     * @param graphics      gui graphics helper
      * @param x             the x position of the rectangle
      * @param y             the y position of the rectangle
      * @param z             the z position of the rectangle
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java:73]: The `ResourceLocation` is ignored

File: common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java [hunk 4]
```diff
@@ -72,7 +73,11 @@ public static void drawRectWithTexture(PoseStack pose, double x, double y, float
      * @param textureWidth  the width of the texture
      * @param textureHeight the height of the texture
      */
-    public static void drawRectWithTexture(PoseStack pose, double x, double y, double z, float u, float v, int width, int height, float textureWidth, float textureHeight) {
+    public static void drawRectWithTexture(ResourceLocation location, GuiGraphics graphics, double x, double y, double z, float u, float v, int width, int height, float textureWidth, float textureHeight) {
+        drawRectWithTexture(location, graphics.pose(), x, y, z, u, v, width, height, textureWidth, textureHeight);
+    }
+
+    public static void drawRectWithTexture(ResourceLocation location, PoseStack pose, double x, double y, double z, float u, float v, int width, int height, float textureWidth, float textureHeight) {
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java:85]: `location` ignored.

File: common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java [hunk 5]
```diff
@@ -91,9 +96,9 @@ public static void drawRectWithTexture(PoseStack pose, double x, double y, doubl
         BufferUploader.drawWithShader(buffer.end());
     }
 
-    public static void drawRectWithFullTexture(PoseStack pose, double x, double y, float u, float v, int width, int height) {
+    public static void drawRectWithFullTexture(GuiGraphics graphics, double x, double y, float u, float v, int width, int height) {
         // Gui.blit(pose, (int) x, (int) y, width, height, u, v, width, height, 256, 256);
-        var e = pose.last().pose();
+        var e = graphics.pose().last().pose();
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java:106]: `location` ignored

File: common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java [hunk 6]
```diff
@@ -104,7 +109,11 @@ public static void drawRectWithFullTexture(PoseStack pose, double x, double y, f
         BufferUploader.drawWithShader(buffer.end());
     }
 
-    public static void drawRectWithTexture(PoseStack pose, double x, double y, float u, float v, int width, int height, float textureWidth, float textureHeight, int sourceWidth, int sourceHeight) {
+    public static void drawRectWithTexture(ResourceLocation location, GuiGraphics graphics, double x, double y, float u, float v, int width, int height, float textureWidth, float textureHeight, int sourceWidth, int sourceHeight) {
+        drawRectWithTexture(location, graphics.pose(), x, y, u, v, width, height, textureWidth, textureHeight, sourceWidth, sourceHeight);
+    }
+
+    public static void drawRectWithTexture(ResourceLocation location, PoseStack pose, double x, double y, float u, float v, int width, int height, float textureWidth, float textureHeight, int sourceWidth, int sourceHeight) {
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java:121]: `location` ignored

File: common/src/main/java/com/ultreon/devices/block/entity/renderer/PrinterRenderer.java [hunk 7]
```diff
@@ -138,7 +139,7 @@ public void render(PrinterBlockEntity blockEntity, float partialTick, @NotNull P
               //  pose.pushPose();
              //   pose.scale(0.1f, -0.1f, 0.1f);
                 pose.scale(0.010416667f, -0.010416667f, 0.010416667f);
-                Minecraft.getInstance().font.drawInBatch(Integer.toString(blockEntity.getPaperCount()), Minecraft.getInstance().font.width("00")-Minecraft.getInstance().font.width(Integer.toString(blockEntity.getPaperCount())), 0, Color.WHITE.getRGB(), false, pose.last().pose(), bufferSource, false, 0x00000000, packedLight);
+                Minecraft.getInstance().font.drawInBatch(Integer.toString(blockEntity.getPaperCount()), Minecraft.getInstance().font.width("00")-Minecraft.getInstance().font.width(Integer.toString(blockEntity.getPaperCount())), 0, Color.WHITE.getRGB(), false, pose.last().pose(), bufferSource, Font.DisplayMode.NORMAL, 0x00000000, packedLight);
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/block/entity/renderer/PrinterRenderer.java:142]: Text displays in wrong position, not sure if present in 1.19.3

File: common/src/main/java/com/ultreon/devices/core/Laptop.java [hunk 8]
```diff
@@ -793,9 +794,8 @@ public boolean mouseScrolled(double mouseX, double mouseY, double delta) {
         return true;
     }
 
-    @Override
-    public void renderComponentTooltip(@NotNull PoseStack pose, @NotNull List<Component> tooltips, int x, int y) {
-        super.renderComponentTooltip(pose, tooltips, x, y);
+    public void renderComponentTooltip(@NotNull GuiGraphics graphics, @NotNull List<Component> tooltips, int x, int y) {
+        graphics.renderComponentTooltip(minecraft.font, tooltips, x, y);
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/core/Laptop.java:798]: Not sure if we should keep this around

File: common/src/main/java/com/ultreon/devices/programs/PixelPainterApp.java [hunk 9]
```diff
@@ -188,10 +188,10 @@ public void render(PoseStack pose, Picture picture, GuiComponent gui, Minecraft
...
         listPictures.setListItemRenderer(new ListItemRenderer<>(20) {
             @Override
-            public void render(PoseStack pose, Picture picture, GuiComponent gui, Minecraft mc, int x, int y, int width, int height, boolean selected) {
-                Gui.fill(pose, x, y, x + width, y + height, selected ? ITEM_SELECTED.getRGB() : ITEM_BACKGROUND.getRGB());
-                mc.font.draw(pose, picture.getName(), x + 2, y + 2, Color.WHITE.getRGB());
-                mc.font.draw(pose, picture.getAuthor(), x + 2, y + 11, AUTHOR_TEXT.getRGB());
+            public void render(GuiGraphics graphics, Picture picture, Minecraft mc, int x, int y, int width, int height, boolean selected) {
+                graphics.fill(x, y, x + width, y + height, selected ? ITEM_SELECTED.getRGB() : ITEM_BACKGROUND.getRGB());
+                graphics.drawString(mc.font, picture.getName(), x + 2, y + 2, Color.WHITE.getRGB());
+                graphics.drawString(mc.font, picture.getAuthor(), x + 2, y + 11, AUTHOR_TEXT.getRGB());
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/programs/PixelPainterApp.java]: Text should have no shadow.

File: common/src/main/java/com/ultreon/devices/programs/auction/MineBayApp.java [hunk 10]
```diff
@@ -145,26 +145,26 @@ public void init(@Nullable CompoundTag intent) {
...
+                    graphics.pose().scale(0.666f, 0.666f, 0);
+                    graphics.drawString(mc.font, e.getStack().getDisplayName(), 0, 0, Color.WHITE.getRGB());
+                    graphics.drawString(mc.font, TimeUtil.getTotalRealTime(e.getTimeLeft()), 0, 11, Color.LIGHT_GRAY.getRGB());
                 }
-                pose.popPose();
+                graphics.pose().popPose();
 
                 String price = "$" + e.getPrice();
-                mc.font.draw(pose, price, x - mc.font.width(price) + width - 5, y + 6, Color.YELLOW.getRGB());
+                graphics.drawString(mc.font, price, x - mc.font.width(price) + width - 5, y + 6, Color.YELLOW.getRGB());
```
Comment (Jab125) [inline common/src/main/java/com/ultreon/devices/programs/auction/MineBayApp.java]: Text should have no shadow

**Review by Jab125 — COMMENTED** [review]
Quick self review:
- A few `Font#draw` calls were accidentally translated to `GuiGraphics#drawString` without `shadow=false`

- `RenderUtil` needs major reworking in a future update, since it contains 90% legacy code. `ResourceLocation` in the rect drawing methods is ignored.

- Item rendering is slightly broken.

- Printer's text renders in the wrong location.

Most methods now take `GuiGraphics` instead of `PoseStack`.

Edit: And the fabric datagen helper fails to compile, which is why the build fails.

File: common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java (hunk 4 above)
Comment (Jab125) [reply common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java:85]: Will be resolved in the future

File: common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java (hunk 6 above)
Comment (Jab125) [reply common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java:121]: Will be resolved in the future

File: common/src/main/java/com/ultreon/devices/core/Laptop.java (hunk 8 above)
Comment (Jab125) [reply common/src/main/java/com/ultreon/devices/core/Laptop.java:798]: Will be resolved in the future

File: common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java (hunk 3 above)
Comment (Jab125) [reply common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java:73]: Will be resolved in the future

File: common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java (hunk 5 above)
Comment (Jab125) [reply common/src/main/java/com/ultreon/devices/api/utils/RenderUtil.java:106]: Will be resolved in the future

File: common/src/main/java/com/ultreon/devices/block/entity/renderer/PrinterRenderer.java (hunk 7 above)
Comment (Jab125) [reply common/src/main/java/com/ultreon/devices/block/entity/renderer/PrinterRenderer.java:142]: Will be addressed in a future PR fixing the router, printer and paper.

**Review by QboiDev — APPROVED** [review]
It's a bit glitchy, but it's definitely a port to 1.20.1.

**Pull request merged**